src/Controllers/RegistrationController.php

```diff
@@ -0,0 +1,268 @@
+<?php
+
+namespace Engelsystem\Controllers;
+
+use Carbon\Carbon;
+use Engelsystem\Config\Config;
+use Engelsystem\Database\Db;
+use Engelsystem\Helpers\Authenticator;
+use Engelsystem\Http\Request;
+use Engelsystem\Http\Response;
+use Engelsystem\Mail\EngelsystemMailer;
+use Engelsystem\Models\User\Contact;
+use Engelsystem\Models\User\PasswordReset;
+use Engelsystem\Models\User\PersonalData;
+use Engelsystem\Models\User\Settings;
+use Engelsystem\Models\User\State;
+use Engelsystem\Models\User\User;
+use Illuminate\Support\Arr;
+use Illuminate\Support\Collection;
+use Parsedown;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+
+class RegistrationController extends BaseController
+{
+    /** @var Response */
+    protected $response;
+
+    /** @var SessionInterface */
+    protected $session;
+
+    /** @var Authenticator */
+    protected $auth;
+
+    /** @var EngelsystemMailer */
+    protected $mail;
+
+    /** @var LoggerInterface */
+    protected $log;
+
+    /** @var Config */
+    protected $config;
+
+    /** @var array */
+    protected $permissions = [
+        'register'      => 'register',
+        'postRegister'  => 'register',
+    ];
+
+    /**
+     * @param Response $response
+     * @param SessionInterface $session
+     * @param Authenticator $auth
+     * @param EngelsystemMailer $mail
+     * @param LoggerInterface $log
+     * @param Config $config
+     */
+    public function __construct(
+        Response $response,
+        SessionInterface $session,
+        Authenticator $auth,
+        EngelsystemMailer $mail,
+        LoggerInterface $log,
+        Config $config
+    ) {
+        $this->response = $response;
+        $this->session = $session;
+        $this->auth = $auth;
+        $this->mail = $mail;
+        $this->log = $log;
+        $this->config = $config;
+    }
+
+    /**
+     * @return Response
+     */
+    public function register(): Response
+    {
+        return $this->showForm();
+    }
+
+    /**
+     * @param Request $request
+     * @return Response
+     */
+    public function postRegister(Request $request): Response
+    {
+        $isPending = $this->config->get('enable_pending_registrations') && $this->auth->user() != null;
+
+        if (!($isPending || $this->config->get('registration_enabled'))) {
+            error(__('registration.disabled'));
+
+            return $this->response->redirectTo('/');
+        }
+
+        $rules = [
+            'login'     => 'required|between:1:23|regex:/^([\p{L}\p{N}\-_. ]+)$/ui',
+            'email'     => 'required|email',
+            'mobile'    => 'optional|phone',
+        ];
+
+        if ($this->config->get('enable_tshirt_size')) {
+            $rules = array_merge($rules, [
```

Comment:
Why is `array_merge` used here? A simple `$rules[] = ...` does the same.

Same for all following.